### PR TITLE
#163653690 Fix cancellation percentage accuracy

### PR DESCRIPTION
**Pull request by RonKbS**
#### What does this PR do?
The PR increases the number of decimal places of the cancellation percentage.
It also adjusts the function used to return the percentage to make it a member of the class in which it had been placed.
In regards to the assertion added in the associated tests file, the line that returns the percentage is not covered and the assertion resolves this.

#### Description of Task to be completed?
Enable a more accurate figure to be returned when running analytics rations, more especially cancellationsPercentage.

#### How should this be manually tested?
Pull the branch and create an odd number of events with a valid calender_id corresponding to a room_id in the local database.
Run the mutation below with corresponding dates for the events created.

``` 
query {
  analyticsRatios (startDate: "Jul 05 2018", endDate: "Jul 16 2018") {
    cancellationsPercentage
    cancellations
    appBookings
    appBookingsPercentage
  }
}
```

The response should be similar to that shown in the screenshots.

#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
[#163653690](https://www.pivotaltracker.com/story/show/163653690)

#### Screenshots (if appropriate)
Before
screen shot 2019-02-01 at 2 42 46 pm

After making changes
screenshot 2019-02-07 at 20 40 42

### CODE REVIEW

File: helpers/calendar/ratios_and_utilization.py [hunk 1]
```diff
@@ -149,14 +149,14 @@ def map_results_to_ratio_class(self,
...
-    def percentage_formater(portion, total):
+    def percentage_formater(self, portion, total):
         """ Calculates the percentage of the entered portion to the total and returns it
          :params
             - portion, total
         """
         try:
             percentage = (portion/total) * 100
-            return round(percentage, 1)
+            return round(percentage, 2)
```
Comment (nakatuddesuzan) [inline helpers/calendar/ratios_and_utilization.py]: Hey @RonKbS, Nice work, I have a small concern, the task is requiring us to be as accurate as possible, Is rounding off to two decimal places the best we can go for or we can go ahead and round off to 3 or 4. Or you could include clear screenshots that show the differences between those two results other than showing the difference in the decimal places. This will show more about the accuracy of 2 decimal places.

**Comment by kampkelly** [comment]
Good work @RonKbS but please look into the concern raised by @nakatuddesuzan.
The percentage is meant to be as accurate as possible.

File: helpers/calendar/ratios_and_utilization.py (hunk 1 above)
Comment (RonKbS) [reply helpers/calendar/ratios_and_utilization.py]: I have gone with removing the removing the round function as recommended by the TC @nakatuddesuzan , @kampkelly and @caveinn.

Tests are failing now because I have included a check for calender_id to make sure data is not corrupted by creating rooms with similar calender_ids. This is because the id is used in analytics to retrieve data for rooms

Comment (bevkololi) [reply helpers/calendar/ratios_and_utilization.py]: Hey @RonKbS, isn't this out of scope?

Comment (RonKbS) [reply helpers/calendar/ratios_and_utilization.py]: I'm seeing it as being part of 'increasing the accuracy of the cancellation percentage'.

The story's description also refers to bookings. I, therefore, thought to add the calender_id check as another means of improving the accuracy because in my case, for example, the percentages returned on running the analytics were similar. This was because the rooms I created had the same calender_id.

File: api/room/schema.py [hunk 2]
```diff
@@ -90,9 +90,21 @@ def mutate(self, info, office_id, **kwargs):
...
         query = Room.get_query(info)
         exact_query = room_join_office(query)
+
+        # prevent analytics data being corrupted by
+        # checking that rooms created have unique calendar ids
+        room_with_same_calender_id = query.filter(
+            RoomModel.state == "active",
+            RoomModel.calendar_id == kwargs['calendar_id']
+        )
+        if room_with_same_calender_id.count() > 0:
```
Comment (Bonifase) [inline api/room/schema.py]: Why do we have to count, can't we just check without counting?

File: fixtures/room/create_room_fixtures.py [hunk 3]
```diff
@@ -164,7 +164,7 @@
     mutation {
         createRoom(
             name: "Kigali", roomType: "Meeting", capacity: 6, floorId: 4, officeId: 1,
-            calendarId:"andela.com_38363233383232303439@resource.calendar.google.com",
+            calendarId:"232k",
```
Comment (Bonifase) [inline fixtures/room/create_room_fixtures.py]: Why was this changed?

Comment (Bonifase) [reply fixtures/room/create_room_fixtures.py]: I would suggest not to round off the percentage values, let the front end team decide on the number of decimal places they'd like to render.

File: api/room/schema.py (hunk 2 above)
Comment (RonKbS) [reply api/room/schema.py]: An object is returned whether it is a positive or negative result as regards the if-check.
.count is, however, able to obtain a negative result on the query result that is run in this case and hence it is employed here.

File: fixtures/room/create_room_fixtures.py (hunk 3 above)
Comment (RonKbS) [reply fixtures/room/create_room_fixtures.py]: The test that used this fixture was hitting the check for existing calender_ids and stopping at that point, returning the error raised by that check.
This change ensures it moves past it and does what is required by the test.

File: helpers/calendar/ratios_and_utilization.py [hunk 4]
```diff
@@ -149,14 +149,14 @@ def map_results_to_ratio_class(self,
...
-    def percentage_formater(portion, total):
+    def percentage_formater(self, portion, total):
         """ Calculates the percentage of the entered portion to the total and returns it
          :params
             - portion, total
         """
         try:
             percentage = (portion/total) * 100
-            return round(percentage, 1)
+            return percentage
```
Comment (RonKbS) [inline helpers/calendar/ratios_and_utilization.py]: > I would suggest not to round off the percentage values as shown in the screenshots, let the front end team decide on the number of decimal places they'd like to render.

The rounding off has been removed but the snapshot has not been updated @Bonifase.
Let me update it now.

**Comment by dnuwa** [comment]
good work @RonKbS I've tested it out and it runs fine.

**Comment by RonKbS** [comment]
> This is quality work @RonKbS, kudos! Please however raise a separate PR for the analytics fix. We'll losing valuable commit history if I merge this change a long with this PR.

Thank you @joshuaocero.

@dnuwa was recently given a task that also implements this `https://github.com/andela/mrm_api/pull/246/files` so I will go ahead and remove it from the PR.
